### PR TITLE
Ce eb mark item purchased

### PR DESCRIPTION
**Pull request by ClaudiaStar**
#11 As a user, I want to mark an item already on my shopping list as purchased.

AC:

- User is able to tap a checkbox or similar UI element to mark an item in the list as purchased

- Item should be shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day)

In other words - allow user to mark item as purchased, and keep the item marked for 24 hours. After 24 hours item should become unmarked but should remain on the list. 

:) :) :)

### CODE REVIEW

**Comment by gunderodd** [comment]
I made a lot of small changes to our PR. I'll try and summarize them here so that they make sense:

**1.** Fixed our tests. 3 were failing. 2 were fixed by using other team’s updated tests. 1 was fixed by adding a function to add a random input instead of “marshmallows” every time, which was outdated due to the new functionality that doesn’t allow the same input twice. This is the "Resets input field after addItem button click" test.

**2.** There are 3 errors that I discovered in both of our branches. I patched 2 of them. Each error was the familiar `FirebaseError: Function CollectionReference.doc() requires its first argument to be of type non-empty string, but it was: null` message. 1. If the user pressed "add item" when nothing was in the input field. 2. If the user pressed "join list" when nothing was in that input field. 3. _This one I couldn't figure out_: If the user presses "Delete Token" when any items are already in their list. 

**3.** Odds and ends. Deleted outdated CSS. Dealt with a react warning about using `<p>` by changing to `<React.Fragment>` on the join page. Also updated a few files to match the other team’s code to try to make the merge easier.

File: cypress/integration/add_item.js [hunk 1]
```diff
@@ -2,24 +2,37 @@ describe("Add Item To List", function() {
...
 
+
+  // This test was not passing, because of a conflict with our new functionality
+  // that handles duplicate items. I added a function to generate a random item,
+  // instead of "marshmallows" every time, so now the test can work with a new
+  // item on every run - Ethan
   it("Resets input field after addItem button click", function() {
     cy.visit("/add");
-    cy.get(".inputField").type("marshmallows");
+    cy.get(".inputField").type(randomInput());
```
Comment (mindyzwan) [inline cypress/integration/add_item.js]: This looks good! Better than our solution - we just deleted it 😂 

One recommendation is to move it above all of the test logic though (line 23). In other situations I've seen helper methods like this extracted fully out of the test suite so it's up on line 1, but not sure if that's best here or not. 🤷‍♀

File: cypress/integration/add_item.js [hunk 2]
```diff
@@ -2,24 +2,37 @@ describe("Add Item To List", function() {
   beforeEach(function() {
     window.localStorage.setItem("uniqueToken", "token1234");
   });
-
```
Comment (mindyzwan) [inline cypress/integration/add_item.js]: Is there a reason you're deleting all of the empty lines between tests?

File: cypress/integration/fetch_items.js [hunk 3]
```diff
@@ -1,37 +1,48 @@
-describe("Show list of items", function() {
```
Comment (mindyzwan) [inline cypress/integration/fetch_items.js]: Might not be a bad idea to do a separate pull request after we merge and switch all of our tests to using single instead of double quotes for consistency :)

File: src/App.css [hunk 4]
```diff
@@ -93,6 +93,26 @@ a:visited {
...
+  list-style-type: none;
+  padding: 10px;
+  margin: 10px;
+  background-color: #ff165d;
+  border-radius: 8%;
+  text-decoration: line-through;
+  cursor: pointer;
+  opacity: 80%;
+}
+
```
Comment (mindyzwan) [inline src/App.css]: bitmoji

Awesome - I like it! 👏

File: src/component/FetchItems.js [hunk 5]
```diff
@@ -7,17 +7,61 @@ import DeleteToken from './DeleteToken';
 const FetchItems = ({ token, setToken, firestore }) => {
   const [empty, setEmpty] = useState(true);
 
+  // stores the number of milliseconds elapsed since January 1, 1970
+  const todayInMs = Date.now();
```
Comment (mindyzwan) [inline src/component/FetchItems.js]: We're using dayjs in our code - might be something to consider here too! Super handy for dealing with date/time differences!

http://zetcode.com/javascript/dayjs/

File: src/component/JoinList.js [hunk 6]
```diff
@@ -2,31 +2,32 @@ import React, { useState } from 'react';
...
+        .get()
+        .then(doc => {
+          if (doc.exists) {
+            addToLS(token);
+            setRedirect(true);
+          } else {
+            alert('Enter a valid share code and try again.');
+          }
+        });
+    }
```
Comment (mindyzwan) [inline src/component/JoinList.js]: I think this is fine! Thanks, Ethan!

File: src/component/JoinList.js [hunk 7]
```diff
@@ -70,7 +73,7 @@ const JoinList = ({ firestore }) => {
       <Link to="/" className="button-link" id="newListButton">
         Start a New List
       </Link>
-    </p>
+    </React.Fragment>
```
Comment (mindyzwan) [inline src/component/JoinList.js]: Just a heads up - I think we did a paragraph here on advice from the mentors. Might not still apply though - I thought I saw somewhere you were saying something about it causing a warning?

File: src/lib/firebase.js [hunk 8]
```diff
@@ -12,9 +11,9 @@ const firebaseConfig = {
   storageBucket: 'tcl-2-smart-shopping-list.appspot.com',
   messagingSenderId: '538706887514',
   appId: '1:538706887514:web:30b4235c1e8aa5fa19a963',
+  timestampsinSnapshots: true,
 };
 
 const fb = firebase.initializeApp(firebaseConfig);
 const db = fb.firestore();
-
```
Comment (mindyzwan) [inline src/lib/firebase.js]: Is there a reason you're deleting lines here?

File: src/component/AddItem.js [hunk 9]
```diff
@@ -19,6 +20,8 @@ const AddItem = ({ firestore }) => {
 
   //   Write item to Firebase setting uniqueToken as document name
   const addItem = (normalizedName, numberOfDays) => {
+    let nextPurchase = new Date();
+    nextPurchase.setDate(today.getDate() + parseInt(numberOfDays));
```
Comment (segdeha) [inline src/component/AddItem.js]: Wait...I saw this in the other PR, too. Was this branch based on that one? That's important info to note in the PR description when it's the case. Thanks!

File: src/component/AddItem.js [hunk 10]
```diff
@@ -81,7 +88,13 @@ const AddItem = ({ firestore }) => {
   const handleSubmit = event => {
     event.preventDefault();
     let normalizedName = normalizeName(name);
-    addItem(normalizedName, numberOfDays);
+    // This conditional allows the add button to be accidentally clicked
+    // while the form is empty without throwing an error
+    if (normalizedName === '') {
```
Comment (segdeha) [inline src/component/AddItem.js]: Hmm. This code should have been in `master` already when you created this branch, right? This is a change from last week, if I recall correctly. I'm a bit confused about the state of this branch 😕

File: src/component/FetchItems.js [hunk 11]
```diff
@@ -7,17 +7,61 @@ import DeleteToken from './DeleteToken';
...
+    event.preventDefault();
+    itemsDocRef
+      .doc(itemId)
+      .get()
+      .then(updateDatabase);
+  };
+
+  const updateDatabase = data => {
+    itemsDocRef.doc(data.id).update({
+      purchasedWithinLastDay: true,
```
Comment (segdeha) [inline src/component/FetchItems.js]: As [I mentioned in Slack](https://the-collab-lab.slack.com/archives/CPPP4NKE3/p1576435576009600?thread_ts=1576435127.009500&cid=CPPP4NKE3), we should compute this value when rendering the list of items rather than storing it in the database.

It is generally good practice to calculate what you can rather than storing it, but in this case the data itself—specifically `purchasedWithinLastDay `—will quickly be out of date. Keeping it up-to-date is possible, but not easy and would be error prone/one more moving part that could have bugs.

File: src/component/JoinList.js [hunk 12]
```diff
@@ -2,31 +2,32 @@ import React, { useState } from 'react';
...
+    if (token === null) {
+      return;
+    } else {
+      firestore
+        .collection('lists')
+        .doc(token)
+        .get()
+        .then(doc => {
+          if (doc.exists) {
+            addToLS(token);
```
Comment (segdeha) [inline src/component/JoinList.js]: Could we change this function name to something more descriptive? What does "LS" stand for?

File: cypress/integration/add_item.js (hunk 2 above)
Comment (ClaudiaStar) [reply cypress/integration/add_item.js]: @mindyzwan I did not reaIize we had deleted empty lines, I can add them back it :)

File: cypress/integration/add_item.js (hunk 1 above)
Comment (ClaudiaStar) [reply cypress/integration/add_item.js]: That seems like a good idea!

File: cypress/integration/fetch_items.js (hunk 3 above)
Comment (ClaudiaStar) [reply cypress/integration/fetch_items.js]: I think the prettier plugin will do that for us. Is that right @alejandronanez?

File: src/App.css (hunk 4 above)
Comment (ClaudiaStar) [reply src/App.css]: Thanks!

File: src/component/FetchItems.js (hunk 5 above)
Comment (ClaudiaStar) [reply src/component/FetchItems.js]: I tried using it at first and it was giving me some trouble. But I'll try it one more time. :)

Comment (ClaudiaStar) [reply src/component/FetchItems.js]: I think this switch will be easier to take care of once we merge our branch since you have some useful logic in calculations.js.

File: src/lib/firebase.js (hunk 8 above)
Comment (ClaudiaStar) [reply src/lib/firebase.js]: We did not mean to delete any lines. Since Ethan and I got started working on our branch after you guys had already made some good changes, we decided to copy and paste all your new code into our branch. That was not a good decision and that is why our branch is all over the place. Sorry for the confusion! We will keep your code when we merge.

File: cypress/integration/add_item.js [hunk 13]
```diff
@@ -2,24 +2,37 @@ describe("Add Item To List", function() {
...
+  // item on every run - Ethan
   it("Resets input field after addItem button click", function() {
     cy.visit("/add");
-    cy.get(".inputField").type("marshmallows");
+    cy.get(".inputField").type(randomInput());
+
+    // the function to create a random input
+    function randomInput() {
+      var text = '';
+      var letterSelection = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
```
[the remark this answers is not in the document]
Comment (ClaudiaStar) [reply cypress/integration/add_item.js]: Good call! I added some of that in there.

File: src/component/AddItem.js (hunk 9 above)
Comment (ClaudiaStar) [reply src/component/AddItem.js]: Ethan and I made a not so good decision earlier this week. Because we got started on our branch a bit late, we decided to copy and paste the changes that the other team had made to their branch thus far into our branch. This was in hopes of using what they had already created in regards to adding new info to the DB and trying to minimize merge conflicts in the future. After messing around with the code, we realized we didn't need much of what they had done after all... live and learn. Sorry for the confusion!

File: src/component/AddItem.js (hunk 10 above)
Comment (ClaudiaStar) [reply src/component/AddItem.js]: See my response above 😕

File: src/component/FetchItems.js (hunk 11 above)
Comment (ClaudiaStar) [reply src/component/FetchItems.js]: @segdeha I realized we don't even need the purchasedWithinLastDay value. We can simply use the time difference logic to update the class names. So I removed that altogether!

File: src/component/JoinList.js (hunk 12 above)
Comment (ClaudiaStar) [reply src/component/JoinList.js]: Changed this to addToLocalStorage.

File: src/component/AddItem.js (hunk 9 above)
Comment (segdeha) [reply src/component/AddItem.js]: All good! Live and learn 🙂

File: src/component/FetchItems.js (hunk 11 above)
Comment (segdeha) [reply src/component/FetchItems.js]: Perfect!

File: src/component/JoinList.js (hunk 12 above)
Comment (segdeha) [reply src/component/JoinList.js]: Excellent, thank you!

File: src/component/JoinList.js (hunk 7 above)
Comment (segdeha) [reply src/component/JoinList.js]: 🤔Nesting a `<form>` inside a `<p>` is invalid HTML. In general, if a tag isn't needed for semantic reasons, it's better to omit it
